File: gateway-service/internal/models/monitoring/prometheus.go
package monitoring

import (
	"context"
	"fmt"
	"io"
	"log/slog"
	"net"
	"net/http"
	"strconv"
	"time"

	monpb "gateway-service/internal/gen/proto/go/vartrack/v1/models/monitoring"
	"gateway-service/internal/protoutil"

	otelprom "go.opentelemetry.io/otel/exporters/prometheus"
	"go.opentelemetry.io/otel/metric"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/collectors"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/client_golang/prometheus/push"
)

// PrometheusBackend exposes a /metrics scrape endpoint and optionally pushes
// to a Pushgateway.
//
// The OTel→Prometheus bridge exporter is used so every AppMetrics instrument
// (counter, histogram, gauge) is automatically visible at the scrape endpoint —
// no manual prometheus.Register calls required anywhere in the app.
//
// Implements:
//   - Backend            (Name / Ping / Shutdown)
//   - meterProviderBackend  (MeterProvider)
type PrometheusBackend struct {
	cfg      *monpb.PrometheusConfig
	registry *prometheus.Registry
	provider *sdkmetric.MeterProvider
	server   *http.Server
	pusher   *push.Pusher
	pushStop chan struct{}
}

// NewPrometheusBackend creates the Prometheus registry, OTel MeterProvider,
// and starts the HTTP exposition server.
// Returns a disabled no-op backend (not an error) when enabled: false.
func NewPrometheusBackend(ctx context.Context, cfg *monpb.PrometheusConfig) (*PrometheusBackend, error) {
	b := &PrometheusBackend{
		cfg:      cfg,
		pushStop: make(chan struct{}),
	}

	if !cfg.GetEnabled() {
		slog.Info("prometheus: disabled")
		return b, nil
	}

	// Dedicated registry — never collides with the global default registerer.
	reg := prometheus.NewRegistry()
	reg.MustRegister(
		// MetricsAll captures all /sched, /gc, /memory etc. runtime metrics.
		collectors.NewGoCollector(
			collectors.WithGoCollectorRuntimeMetrics(collectors.MetricsAll),
		),
		collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}),
	)
	b.registry = reg

	res, err := buildOTelResource(ctx, cfg.GetJobName(), "", "", cfg.GetExternalLabels())
	if err != nil {
		return nil, fmt.Errorf("prometheus: resource: %w", err)
	}

	// OTel→Prometheus bridge: registers OTel instruments into our dedicated registry.
	promExp, err := otelprom.New(otelprom.WithRegisterer(reg))
	if err != nil {
		return nil, fmt.Errorf("prometheus: OTel exporter: %w", err)
	}

	b.provider = sdkmetric.NewMeterProvider(
		sdkmetric.WithReader(promExp),
		sdkmetric.WithResource(res),
	)

	if port := cfg.GetPort(); port > 0 {
		if err := b.startServer(cfg); err != nil {
			return nil, fmt.Errorf("prometheus: start server: %w", err)
		}
	}

	if cfg.GetPushEnabled() {
		if err := b.startPushLoop(cfg); err != nil {
			return nil, fmt.Errorf("prometheus: start push loop: %w", err)
		}
	}

	slog.Info("prometheus: started",
		"port", cfg.GetPort(),
		"path", cfg.GetMetricsPath(),
		"push_enabled", cfg.GetPushEnabled(),
	)
	return b, nil
}

func (b *PrometheusBackend) Name() string { return "prometheus" }

// MeterProvider satisfies meterProviderBackend.
// monitoring.Init() installs this as the global OTel MeterProvider.
func (b *PrometheusBackend) MeterProvider() metric.MeterProvider {
	if b.provider == nil {
		return nil
	}
	return b.provider
}

// Ping does a live GET /metrics against the local exposition server.
// A failure means the server goroutine has died — worth logging, not worth
// stopping traffic over.
func (b *PrometheusBackend) Ping(ctx context.Context) error {
	if b.server == nil {
		return nil
	}

	metricsPath := b.cfg.GetMetricsPath()
	if metricsPath == "" {
		metricsPath = "/metrics"
	}
	scheme := "http"
	if b.cfg.GetEnableTls() {
		scheme = "https"
	}
	url := scheme + "://" + b.server.Addr + metricsPath

	reqCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, url, nil)
	if err != nil {
		return fmt.Errorf("prometheus ping: %w", err)
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("prometheus ping GET %s: %w", url, err)
	}
	io.Copy(io.Discard, resp.Body)
	resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("prometheus ping: %s returned HTTP %d", url, resp.StatusCode)
	}
	return nil
}

// Shutdown stops the Pushgateway loop, the HTTP server, and the MeterProvider.
func (b *PrometheusBackend) Shutdown(ctx context.Context) error {
	var errs []error

	if b.provider != nil {
		if err := b.provider.Shutdown(ctx); err != nil {
			errs = append(errs, err)
		}
	}

	// Signal push loop goroutine to stop.
	if b.pushStop != nil {
		select {
		case <-b.pushStop: // already closed
		default:
			close(b.pushStop)
		}
	}

	if b.server != nil {
		if err := b.server.Shutdown(ctx); err != nil {
			errs = append(errs, err)
		}
	}

	if len(errs) > 0 {
		return fmt.Errorf("prometheus shutdown: %v", errs)
	}
	slog.Info("prometheus: shut down")
	return nil
}

// ── HTTP exposition server ────────────────────────────────────────────────────

func (b *PrometheusBackend) startServer(cfg *monpb.PrometheusConfig) error {
	metricsPath := cfg.GetMetricsPath()
	if metricsPath == "" {
		metricsPath = "/metrics"
	}

	handler := promhttp.HandlerFor(b.registry, promhttp.HandlerOpts{
		Registry:          b.registry,
		EnableOpenMetrics: true, // enables content negotiation for OpenMetrics format
	})

	if user := cfg.GetBasicAuthUser(); user != "" {
		handler = basicAuthMiddleware(user, cfg.GetBasicAuthPass(), handler)
	}

	mux := http.NewServeMux()
	mux.Handle(metricsPath, handler)
	// /healthz on the same port lets k8s liveness probes reuse it cheaply.
	mux.HandleFunc("/healthz", func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusOK)
		w.Write([]byte("OK"))
	})

	addr := net.JoinHostPort("", strconv.Itoa(int(cfg.GetPort())))
	srv := &http.Server{
		Addr:              addr,
		Handler:           mux,
		ReadHeaderTimeout: 5 * time.Second,
		ReadTimeout:       10 * time.Second,
		WriteTimeout:      30 * time.Second,
		IdleTimeout:       60 * time.Second,
	}

	if cfg.GetEnableTls() {
		tlsCfg, err := buildBackendTLS(cfg.GetTlsCa(), cfg.GetTlsCert(), cfg.GetTlsKey(), cfg.GetInsecureSkipVerify())
		if err != nil {
			return err
		}
		srv.TLSConfig = tlsCfg
	}

	go func() {
		var serveErr error
		if cfg.GetEnableTls() {
			serveErr = srv.ListenAndServeTLS("", "") // certs already in TLSConfig
		} else {
			serveErr = srv.ListenAndServe()
		}
		if serveErr != nil && serveErr != http.ErrServerClosed {
			slog.Error("prometheus: exposition server error", "error", serveErr)
		}
	}()

	b.server = srv
	slog.Info("prometheus: exposition server started", "addr", addr, "path", metricsPath)
	return nil
}

// ── Pushgateway loop ──────────────────────────────────────────────────────────

func (b *PrometheusBackend) startPushLoop(cfg *monpb.PrometheusConfig) error {
	p := push.New(cfg.GetPushUrl(), cfg.GetJobName()).Gatherer(b.registry)
	for k, v := range cfg.GetExternalLabels() {
		p = p.Grouping(k, v)
	}
	b.pusher = p

	interval := protoutil.DurationOrDefault(cfg.GetPushInterval(), 15*time.Second)

	go func() {
		ticker := time.NewTicker(interval)
		defer ticker.Stop()
		for {
			select {
			case <-b.pushStop:
				return
			case <-ticker.C:
				if err := b.pusher.Push(); err != nil {
					slog.Warn("prometheus: push failed",
						"url", cfg.GetPushUrl(), "error", err)
				}
			}
		}
	}()

	slog.Info("prometheus: push loop started",
		"url", cfg.GetPushUrl(), "interval", interval)
	return nil
}

func basicAuthMiddleware(user, pass string, next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		u, p, ok := r.BasicAuth()
		if !ok || u != user || p != pass {
			w.Header().Set("WWW-Authenticate", `Basic realm="metrics"`)
			http.Error(w, "Unauthorized", http.StatusUnauthorized)
			return
		}
		next.ServeHTTP(w, r)
	})
}

File: orchestrator-service/app/monitoring/prometheus.py
"""
app/monitoring/prometheus.py
─────────────────────────────
Prometheus monitoring backend.

Exposes a /metrics scrape endpoint on a dedicated port and optionally
pushes to a Pushgateway.  Mirrors gateway's models/monitoring/prometheus.go.
"""
from __future__ import annotations

import logging
import threading

from app.monitoring.base import Backend
from app.monitoring._utils import _build_ssl_context
from app.monitoring.config import PrometheusConfig

logger = logging.getLogger(__name__)


class PrometheusBackend(Backend):
    """
    Exposes a Prometheus /metrics scrape endpoint on a dedicated port and
    optionally pushes to a Pushgateway.

    Mirrors gateway's models/monitoring/prometheus.go PrometheusBackend:
      - Dedicated HTTP server on cfg.port (separate from the FastAPI port)
      - OTel→Prometheus bridge (when opentelemetry-sdk is installed)
      - Pushgateway loop (background thread, cfg.push_interval_seconds)
      - Basic auth middleware (cfg.basic_auth_user / basic_auth_pass)
      - TLS server (cfg.enable_tls)
    """

    def __init__(self, cfg: PrometheusConfig, registry) -> None:
        self._cfg      = cfg
        self._registry = registry
        self._server   = None
        self._push_stop = threading.Event()

        if not cfg.enabled:
            logger.info("prometheus backend: disabled")
            return

        if cfg.port > 0:
            self._start_server()

        if cfg.push_enabled:
            self._start_push_loop()

        logger.info(
            "prometheus backend: started port=%d path=%s push=%s",
            cfg.port, cfg.metrics_path, cfg.push_enabled,
        )

    def name(self) -> str:
        return "prometheus"

    def ping(self) -> None:
        """GET /metrics on the local scrape server."""
        if self._server is None:
            return
        import http.client
        host = f"localhost:{self._cfg.port}"
        path = self._cfg.metrics_path or "/metrics"
        try:
            conn = http.client.HTTPConnection(host, timeout=3)
            conn.request("GET", path)
            resp = conn.getresponse()
            resp.read()
            if resp.status != 200:
                raise RuntimeError(f"prometheus ping: HTTP {resp.status}")
        finally:
            conn.close()

    def shutdown(self) -> None:
        self._push_stop.set()
        if self._server is not None:
            self._server.shutdown()
            logger.info("prometheus backend: shut down")

    def _start_server(self) -> None:
        """
        Start a minimal HTTP server serving /metrics and /healthz.
        Mirrors gateway's startServer() with basic auth + TLS support.
        """
        import http.server
        from prometheus_client import CONTENT_TYPE_LATEST

        cfg      = self._cfg
        registry = self._registry

        class _Handler(http.server.BaseHTTPRequestHandler):
            def log_message(self, fmt, *args):
                pass  # silence access log; metrics server is high-frequency

            def _check_auth(self) -> bool:
                if not cfg.basic_auth_user:
                    return True
                import base64
                auth = self.headers.get("Authorization", "")
                if not auth.startswith("Basic "):
                    return False
                decoded = base64.b64decode(auth[6:]).decode()
                user, _, pwd = decoded.partition(":")
                return user == cfg.basic_auth_user and pwd == cfg.basic_auth_pass

            def do_GET(self):
                if not self._check_auth():
                    self.send_response(401)
                    self.send_header("WWW-Authenticate", 'Basic realm="metrics"')
                    self.end_headers()
                    return

                path = cfg.metrics_path or "/metrics"
                if self.path == "/healthz":
                    self.send_response(200)
                    self.end_headers()
                    self.wfile.write(b"OK")
                elif self.path == path:
                    from prometheus_client import generate_latest
                    output = generate_latest(registry)
                    self.send_response(200)
                    self.send_header("Content-Type", CONTENT_TYPE_LATEST)
                    self.end_headers()
                    self.wfile.write(output)
                else:
                    self.send_response(404)
                    self.end_headers()

        server = http.server.HTTPServer(("", cfg.port), _Handler)

        if cfg.enable_tls and (cfg.tls_cert or cfg.tls_ca):
            ctx = _build_ssl_context(
                cfg.tls_ca, cfg.tls_cert, cfg.tls_key, cfg.insecure_skip_verify
            )
            server.socket = ctx.wrap_socket(server.socket, server_side=True)

        t = threading.Thread(target=server.serve_forever, daemon=True, name="prom-scrape-server")
        t.start()
        self._server = server
        logger.info("prometheus scrape server: listening on :%d%s", cfg.port, cfg.metrics_path)

    def _start_push_loop(self) -> None:
        """
        Background thread that pushes metrics to a Pushgateway.
        Mirrors gateway's startPushLoop() with grouping labels.
        """
        cfg      = self._cfg
        registry = self._registry
        stop     = self._push_stop

        def _loop():
            from prometheus_client import push_to_gateway
            interval = cfg.push_interval_seconds or 15.0
            while not stop.wait(interval):
                try:
                    groupings = {k: v for k, v in cfg.external_labels.items()}
                    push_to_gateway(
                        gateway=cfg.push_url,
                        job=cfg.push_job_name or "orchestrator",
                        registry=registry,
                        grouping_key=groupings or None,
                    )
                except Exception:
                    logger.warning(
                        "prometheus push failed url=%s", cfg.push_url, exc_info=True
                    )

        t = threading.Thread(target=_loop, daemon=True, name="prom-push-loop")
        t.start()
        logger.info(
            "prometheus push loop: url=%s interval=%ss",
            cfg.push_url, cfg.push_interval_seconds,
        )
